### PR TITLE
(FACT-2551) Defaulting resolver keys to nil after no resolution

### PR DESCRIPTION
**Pull request by sebastian-miclea**

### CODE REVIEW

**Comment by sebastian-miclea** [comment]
Jenkins please test this on all

File: spec/facter/resolvers/base_resolver_spec.rb [hunk 1]
```diff
@@ -92,6 +92,24 @@ def self.post_resolve(fact_name)
...
+    before do
+      allow(resolver).to receive(:validate_resolution)
+    end
+
+    it 'sets the fact to nil if undefined' do
+      resolver.validate_resolution('unresolved_fact')
+      expect(resolver.resolve('unresolved_fact')).to be_nil
+    end
+
+    it 'does not overwirte values' do
```
Comment (Filipovici-Andrei) [inline spec/facter/resolvers/base_resolver_spec.rb:105]: typo in overwrite

File: lib/facter/resolvers/base_resolver.rb [hunk 2]
```diff
@@ -18,13 +18,21 @@ def self.subscribe_to_manager
       def self.resolve(fact_name)
         @semaphore.synchronize do
           subscribe_to_manager
+
           post_resolve(fact_name)
```
Comment (Filipovici-Andrei) [inline lib/facter/resolvers/base_resolver.rb:22]: Would it be ok to call post_resolve only if @fact_list is empty?

Comment (sebastian-miclea) [reply lib/facter/resolvers/base_resolver.rb:22]: Not really, you loose control on your resolver call if the execution stops in base_resolver. Compared to exceptions that are caught in base_resolver, this is a more common flow that allows you to 'cache' successful resolver calls that have no output.

Comment (IrimieBogdan) [reply lib/facter/resolvers/base_resolver.rb:22]: We can have partial results in @fact_list so we cannot decide in the base resolver if we want to call the resolver or not. We should always call the resolver.

**Review by IrimieBogdan — APPROVED** [review]

**Pull request merged**